---
title: "Ecology of bacteria in Larry's data"
output: html_document
date: "2025-10-9"
---

```{r DiversityFuns}
library('vegan')
library('data.table')
library('magrittr')
library('ape')
library('ggplot2')
library('maaslin3')


source('scripts/preprocessing.R')
source('scripts/diversity.R')

## set a seed here; dmms and dbrda use randomness.
set.seed(42)
```

```{r AmpliconData}
amplicon_dt = fread('sanitized-data/amplicon-matrix.txt.xls', sep = '\t')
amplicon_matrix = amplicon_dt[,-1] %>% as.matrix()
rownames(amplicon_matrix) = amplicon_dt[,V1]

amplicon_matrix = amplicon_matrix / 100

## amplicon data
amplicon_taxonomy_dt = fread('sanitized-data/amplicon-taxonomy.txt.xls', sep = '\t', header=TRUE)
amplicon_taxonomy = amplicon_taxonomy_dt %>% setkey(V1)

```

```{r}
## We don't have the original proportions, but we can still run effective
## microbial richness.
amplicon_emr = EMR(amplicon_matrix)

## Run shapiro test.
amplicon_emr %>% log2() %>% shapiro.test()

## Bray-curtis
amplicon_bray = vegdist(amplicon_matrix %>% t(), method = 'bray')
amplicon_jacc = vegdist(amplicon_matrix %>% t(), method = 'jaccard')

## Centered log-ratio transformation for aitchison distance
amplicon_clr = amplicon_matrix %>%
  t() %>%
  clrtrans(pseudocount = 1e-5) %>%
  t()
amplicon_aitc = vegdist(amplicon_clr %>% t(), method = 'euclidean')
```
# DMMs
Run DMMs and report laplace for all attempted DMM configurations
```{r DMMs}
## Generate DMMs.
if(!exists('amplicon_dmms')) amplicon_dmms = generate.dmms(amplicon_matrix);

amplicon_dmms_samples = amplicon_dmms[[1]]
laplace_scores = amplicon_dmms[[3]] %>% vapply(laplace, numeric(1)) %>% data.frame(laplace=.)
laplace_scores$nDMMs = (1:nrow(laplace_scores))+1

## Plot laplace scores
ggplot(data = laplace_scores, aes(x = nDMMs, y = laplace)) + 
  geom_point() + 
  geom_line() +
  annotate(geom='text', x = amplicon_dmms[[2]] + 1, y = min(laplace_scores$laplace)* (1 + 1e-3), label = '*') +
  theme_bw()
```

```{r Metagenome}
## Read in metagenome taxonomy (in this case, it's important)
metagenome_taxonomy = fread('sanitized-data/metagenome-taxonomy.txt.xls', sep = '\t',header=TRUE)

## Exclude viruses.
exclude_taxa = metagenome_taxonomy[d == 'Viruses',V1]
metagenome_dt = fread('sanitized-data/metagenome-matrix.txt.xls', sep = '\t')
metagenome_dt = metagenome_dt[!(V1 %in% exclude_taxa),]

## Make metagenome proportional abundance matrix
metagenome_matrix = metagenome_dt[,-1] %>% as.matrix()
rownames(metagenome_matrix) = metagenome_dt[,V1]
metagenome_matrix = metagenome_matrix / 100

## Exclude zero-abundance taxa
metagenome_matrix = metagenome_matrix[rowSums(metagenome_matrix) > 0,]

```

```{r}
## We don't have the original proportions, but we can still run effective
## microbial richness.
metagenome_emr = EMR(metagenome_matrix)

## Run shapiro test.
metagenome_emr %>% log2() %>% shapiro.test()

## Bray-curtis
metagenome_bray = vegdist(metagenome_matrix %>% t(), method = 'bray')
metagenome_jacc = vegdist(metagenome_matrix %>% t(), method = 'jaccard')

## Centered log-ratio transformation for aitchison distance
metagenome_clr = metagenome_matrix %>%
  t() %>%
  clrtrans(pseudocount = 1e-5) %>%
  t()
metagenome_aitc = vegdist(metagenome_clr %>% t(), method = 'euclidean')
```

# MetagenomeDMMs
Run DMMs and report laplace for all attempted DMM configurations
```{r MetagenomeDMMs}
## Generate DMMs.
if(!exists('metagenome_dmms')) metagenome_dmms = generate.dmms(metagenome_matrix);

metagenome_dmms_samples = metagenome_dmms[[1]]
laplace_scores = metagenome_dmms[[3]] %>% vapply(laplace, numeric(1)) %>% data.frame(laplace=.)
laplace_scores$nDMMs = (1:nrow(laplace_scores))+1

## Plot laplace scores
ggplot(data = laplace_scores, aes(x = nDMMs, y = laplace)) + 
  geom_point() + 
  geom_line() +
  annotate(geom='text', x = metagenome_dmms[[2]] + 1, y = min(laplace_scores$laplace)* (1 + 1e-3), label = '*') +
  theme_bw()
```

Sanity-check: the metagenome comparison table should be sparse.

```{r}
comparison = data.table(amplicon = amplicon_dmms_samples[names(metagenome_dmms_samples)]
         , metagenome = metagenome_dmms_samples)

comparison[,table(amplicon,metagenome)]

```

```{r}

## Run DMM abundance
amplicon_metadata = fread('sanitized-data/amplicon-metadata.txt.xls', sep = '\t')
amplicon_metadata = amplicon_dmms_samples %>% cbind(amplicon_metadata, DMM = .)

maaslin3_otu_fit = maaslin3(input_data = amplicon_matrix,
                    input_metadata = amplicon_metadata,
                    output = '.tmp.otu.amplicon_dmm_output',
                    formula = '~ DMM',
                    normalization = 'TSS',
                    transform = 'LOG',
                    augment = TRUE,
                    standardize = TRUE,
                    max_significance = 0.1,
                    median_comparison_abundance = TRUE,
                    median_comparison_prevalence = FALSE,
                    max_pngs = 10,
                    cores = 1)


maaslin3_phylum_fit = maaslin3(input_data = agglomerate(amplicon_matrix, amplicon_taxonomy, rank = 'Phylum'),
                    input_metadata = amplicon_metadata,
                    output = '.tmp.phylum.amplicon_dmm_output',
                    formula = '~ DMM',
                    normalization = 'TSS',
                    transform = 'LOG',
                    augment = TRUE,
                    standardize = TRUE,
                    max_significance = 0.1,
                    median_comparison_abundance = TRUE,
                    median_comparison_prevalence = FALSE,
                    max_pngs = 10,
                    cores = 1)


maaslin3_genus_fit = maaslin3(input_data = agglomerate(amplicon_matrix, amplicon_taxonomy, rank = 'Genus'),
                    input_metadata = amplicon_metadata,
                    output = '.tmp.genus.amplicon_dmm_output',
                    formula = '~ DMM',
                    normalization = 'TSS',
                    transform = 'LOG',
                    augment = TRUE,
                    standardize = TRUE,
                    max_significance = 0.1,
                    median_comparison_abundance = TRUE,
                    median_comparison_prevalence = FALSE,
                    max_pngs = 10,
                    cores = 1)


maaslin3_family_fit = maaslin3(input_data = agglomerate(amplicon_matrix, amplicon_taxonomy, rank = 'Family'),
                    input_metadata = amplicon_metadata,
                    output = '.tmp.family.amplicon_dmm_output',
                    formula = '~ DMM',
                    normalization = 'TSS',
                    transform = 'LOG',
                    augment = TRUE,
                    standardize = TRUE,
                    max_significance = 0.1,
                    median_comparison_abundance = TRUE,
                    median_comparison_prevalence = FALSE,
                    max_pngs = 10,
                    cores = 1)

```

```{r}
maaslin3_genus_fit$fit_data_abundance$results %>% dplyr::filter(qval_individual < 0.05)
maaslin3_genus_fit$fit_data_prevalence$results%>% dplyr::filter(qval_individual < 0.05)
```

```{r fig.width = 8.5}
plot_maaslin3 = function(maaslin3_fit)
{
  .extract_res = function(fit_data)
  {
    fit_res = fit_data$results %>% as.data.table()
    
    fit_features = fit_res[qval_individual < 0.05, feature] %>% unique()
    
    fit_res_sig = fit_res[feature %in% fit_features & is.na(error),]
    return(fit_res_sig)
  }
  
  .assign_qval = function(x) case_when(
      x < 0 ~ NA,
      x < 0.05 ~ 'q < 0.05 *',
      x < 0.10 ~ 'q < 0.1 .',
      TRUE ~ 'N.S.'
  )
  
  res_abund = maaslin3_fit$fit_data_abundance %>% .extract_res()
  res_prevl = maaslin3_fit$fit_data_prevalence %>% .extract_res()
  
  res = rbind(res_abund, res_prevl)
  
  ## Clean this up
  res$feature = gsub('Unclassified','Uncl.', res$feature)
  
  ggplot(data = res) +
    geom_bar(aes(x = coef, y = feature, fill = factor(.assign_qval(qval_individual))), stat = 'identity') +
    geom_errorbar(aes( y = feature, xmin = coef - 2*stderr, xmax = coef + 2*stderr), width = 0.5) +
    geom_vline(aes(xintercept = 0), linetype = 'dashed') +
    facet_grid(model~value, space = 'free_y', scales = 'free_y') +
    theme_bw() +
    scale_fill_signif
    
}

plot_maaslin3(maaslin3_genus_fit) +
    ggtitle('Maaslin3 DMM fit: Genus-level Abundance and Prevalence')
```

# Exit
```{r}
save.image(file = 'sanitized-data/diversity.RData')
sessionInfo()
```